### ISSUE TITLE
JWT: Claims based Validation and Routing

### ISSUE THREAD

**Issue by anderius** [nginx/kubernetes-ingress#6829]
## User Story

As an infrastructure team managing secure Kubernetes applications, I can use JWT claims in NGINX Ingress Controller policies and routing rules so that I can authorize and route requests without modifying the application.

- Assumes NGINX Ingress Controller is in the data path.
- Validates the JWT before using any claims.
- Checks required claims before forwarding traffic.
- Allows routes to make decisions based on validated claim values.
- Rejects unauthorized requests before they reach the application.
## Feature Overview

This document describes the requirements for supporting JWT claims-based authorization and routing in NGINX Ingress Controller.

The feature extends the existing JWT Policy so that users can require specific claims and use validated claims in `Ingress, VirtualServer` and `VirtualServerRoute` match conditions.

NGINX Ingress controller supports dynamically fetching public keys from IdP for JWT validation.

**Prerequisite: Make the JWT policy work with kind: Ingress. **

## What

An NGINX Ingress Controller configuration layer that:

1. Validates the JWT attached to an incoming request.
2. Checks that required claims are present and contain permitted values.
3. Blocks requests that fail authentication or authorization.
4. Makes validated claims available for route matching.
5. Routes authorized requests to the appropriate upstream.
Note: Existing JWT claim-based rate-limiting functionality must remain backward compatible and requires an enhancement. Any use of JWT claims for rate limiting must clearly distinguish successfully validated claims. Only authenticated claim values can affect a rate-limit counter. Example use-case that should be achievable: https://github.com/nginx/kubernetes-ingress/discussions/10371

## Why

Organizations and Agents commonly use JWT claims to represent roles, tenants, client applications, subscription levels, and permissions.

Supporting these claims at the ingress layer provides:

- **Centralized authorization** without duplicating logic across applications.
- **Claims-aware routing** based on trusted token information.
- **Early rejection** of unauthorized requests.
- **Consistent enforcement** across services.
- **Operational simplicity** through Kubernetes configuration.
## Acceptance Criteria

- A missing, malformed, expired, or invalid JWT is rejected before reaching the upstream.
- A request missing a required claim is rejected.
- A request containing an unauthorized claim value is rejected.
- An authorized request is forwarded with its original body unchanged.
- A validated claim can be used in a route match condition.
- Different claim values can route requests to different upstreams.
- A request that matches no claim-based route follows the configured default route.
- Claims are never trusted before JWT validation succeeds.
- JWT processing and key-retrieval failures fail closed.
- Tokens and sensitive claim values are not logged by default.
## Functional Requirements

- The existing JWT Policy must support required claim validation.
- Claim validation must support:

    - Claim presence.
    - Exact value matching.
    - Negative matching.
    - Regular-expression matching.

- Multiple claim requirements may be configured for a policy.
- All configured requirements must pass before the request is forwarded.
- Validated JWT claims must be supported in `VirtualServer` and `VirtualServerRoute` match conditions.
- Claim-based routing must work alongside existing header, cookie, argument, and request-variable matching.
- Authentication failures must return `HTTP 401`.
- Authorization failures must return `HTTP 403`, or a configurable supported status code.
- Invalid policy or route configuration must be rejected without replacing the last valid configuration.
- Remote key-service failures must block requests when no valid cached keys are available.
- The feature must provide logs and metrics that distinguish authentication, authorization, and key-retrieval failures.
- The feature requires a supported NGINX Plus deployment.
## Non-Functional Requirements

- Logs and metrics must not expose JWTs or create unbounded claim-based labels.
- Existing JWT Policies and routes that do not use claim-based features must remain backward compatible.
## Out of Scope

- Issuing or refreshing JWTs.
- Modifying JWT contents.
- Replacing application-level authorization.
- External authorization services.
- Automatically forwarding all claims to upstream applications.
- Routing based on unvalidated JWT data.
- Support for NGINX Open Source in this milestone (possibly in future)
## Documentation Requirements

- **Configuration Guide:** Required claims and claims-based route matching.
- **Configuration Reference:** Supported claim checks, route variables, defaults, and error responses.
- **Security Guidance:** Fail-closed behavior, logging, key handling, and claim privacy.
## Definition of Done

- Required JWT claims can be configured and enforced.
- Validated JWT claims can be used in route match conditions.
- Unauthorized requests are blocked before reaching an upstream.
- Authorized requests are routed to the expected upstream.
- JWT and key-service failures fail closed.
- Tests cover:

    - Missing or invalid JWT.
    - Required claim missing.
    - Claim allowed.
    - Claim rejected.
    - Claims-based routing.
    - Default routing.
    - Key service unavailable.

- Existing JWT configurations remain backward compatible.
- All documentation requirements are complete.
- A security review confirms that claims are used only after validation and that sensitive token data is not exposed.

Extras:

For users that have a requirement to validate specific claims in a JWT token or to perform routing based on a claim, this will allow the capability to inspect JWT tokens for specific `claims`. By inspecting these claims, customers can further take action on `claims` inside the token, or `claims` missing from a JWT token, giving them powerful routing capabilities based on a JWT token.

A few new items will need to be added to NGINX Ingress controller.
These include:

- we need to add the variable `$jwt_payload` for enhanced capabilities.
- We also need to add the `$jwt_` as supported `variable` options when using `match` and `conditions`
- using `auth_jwt_require` to do `claims` validation/routing.
NGINX Ingress controller can be configured to look for specific `claims` in a JWT token

```
apiVersion: k8s.nginx.org/v1kind: VirtualServermetadata:
  name: cafespec:
  host: cafe.example.com
  tls:
    secret: cafe-secret
  gunzip: on
  upstreams:
  - name: tea
    service: tea-svc
    port: 80
 - name: coffee
   service: coffee-svc
   port: 80
 routes:
 - path: /coffee
    matches:
     - conditions:
       - variable: $jwt_client_id
         value: 05
  action:
    pass: coffee
```

Documentation on [VirtualServer.Routes.path.matches.conditions](https://docs.nginx.com/nginx-ingress-controller/configuration/virtualserver-and-virtualserverroute-resources/#condition)

Prototype policy:

```
apiVersion: k8s.nginx.org/v1kind: Policymetadata:
  name: jwt-policyspec:
  jwt:
    realm: MyProductAPI
    token: $http_token
    jwksURI: http://keycloak.default.svc.cluster.local:8080/realms/jwks-example/protocol/openid-connect/certs
    keyCache: 1h
    auth_require: $jwt_client_id $valid_issuer ### These are claims we can we can require for JWT tokens
```

In the prototype above, we can enforce a number of different claims with `auth_require`.
In the virtualserver resource, wee can use `match` to look for specific `$jwt_client_id` and a defined value and then proxy to the backend application.
We can further enhance this by sending back specific HTTP codes (401 or 403) if a claim is empty or has the incorrect value.

**Comment by jasonwilliams14** [comment]
@anderius I believe this is what you are looking for? This capability in NGINX Ingress controller:

https://github.com/nginxinc/kubernetes-ingress/discussions/4417

**Comment by anderius** [comment]
Thank you, I was not aware of that discussion. Would you prefer to have feature requests there instead of issues?

A few points:

 * The routing is not something I need (just to clarify this request)
 * I could not understand the example, the prototype policy, but if the functionality is like in this feature request, all is good

**Comment by shaun-nx** [comment]
Hi @anderius I've added this issue as a child issues for https://github.com/nginxinc/kubernetes-ingress/issues/6565 since this it likely work that will need to come as part of this effort

**Comment by shaun-nx** [comment]
Hi @anderius just wanted to update you on this.
I originally thought our work to enable Rate Limiting with JWT would overlap with this work.
I don't believe we'll get to this work by release v4.1.0 at this point.
We'll do our best to add this capability in a future release. Once I know which release that will be I will let you know.
